File: config/seeder.go
package config

import (
	"log"

	"backend-sipeka/models"

	"golang.org/x/crypto/bcrypt"
)

// SeedDatabase memastikan akun demo default dan data awal tersedia di database
func SeedDatabase() {
	// Hash password default: "password123"
	hashedPass, err := bcrypt.GenerateFromPassword([]byte("password123"), bcrypt.DefaultCost)
	if err != nil {
		log.Printf("Gagal hash password seeder: %v", err)
		return
	}

	// 1. Akun Kasir Default (1234567890)
	var kasir models.User
	if err := DB.Where("nisn_nip = ?", "1234567890").First(&kasir).Error; err != nil {
		kasir = models.User{
			NisnNip:      "1234567890",
			Name:         "Ahmad Kasir",
			PasswordHash: string(hashedPass),
			Role:         "kasir",
			IsActive:     true,
		}
		if err := DB.Create(&kasir).Error; err != nil {
			log.Printf("Gagal seed akun kasir: %v", err)
		} else {
			log.Println(" -> Kasir Default berhasil dibuat: NISN/NIP 1234567890 | Password: password123")
		}
	}

	// 2. Akun Penitip Barang Default (1122334455)
	var penitip models.User
	if err := DB.Where("nisn_nip = ?", "1122334455").First(&penitip).Error; err != nil {
		penitip = models.User{
			NisnNip:      "1122334455",
			Name:         "Ibu Siti Penitip",
			PasswordHash: string(hashedPass),
			Role:         "penitip",
			IsActive:     true,
		}
		if err := DB.Create(&penitip).Error; err != nil {
			log.Printf("Gagal seed akun penitip: %v", err)
		} else {
			log.Println(" -> Penitip Default berhasil dibuat: NISN/NIP 1122334455 | Password: password123")
		}
	}

	// 3. Akun Admin Default (9999999999)
	var admin models.User
	if err := DB.Where("nisn_nip = ?", "9999999999").First(&admin).Error; err != nil {
		admin = models.User{
			NisnNip:      "9999999999",
			Name:         "Administrator PKK SMKN 8",
			PasswordHash: string(hashedPass),
			Role:         "admin",
			IsActive:     true,
		}
		if err := DB.Create(&admin).Error; err != nil {
			log.Printf("Gagal seed akun admin: %v", err)
		} else {
			log.Println(" -> Admin Default berhasil dibuat: NISN/NIP 9999999999 | Password: password123")
		}
	}

	// 4. Produk Konsinyasi Awal (jika belum ada produk)
	var productCount int64
	DB.Model(&models.Product{}).Count(&productCount)
	if productCount == 0 {
		products := []models.Product{
			{
				PenitipID:    penitip.ID,
				Name:         "Roti Bakar Keju",
				Price:        15000,
				SchoolMargin: 1000,
				Stock:        25,
				IsActive:     true,
			},
			{
				PenitipID:    penitip.ID,
				Name:         "Risoles Mayo Spesial",
				Price:        5000,
				SchoolMargin: 500,
				Stock:        30,
				IsActive:     true,
			},
			{
				PenitipID:    penitip.ID,
				Name:         "Teh Kotak Sosro 250ml",
				Price:        4000,
				SchoolMargin: 500,
				Stock:        50,
				IsActive:     true,
			},
		}
		for i := range products {
			DB.Create(&products[i])
		}
		log.Println(" -> Produk konsinyasi awal berhasil ditambahkan")
	}

	// 5. Dummy Order Pre-Order untuk di-Scan (jika belum ada)
	qrCode := "ORD-PRE-20260923-001"
	var existingOrder models.Order
	if err := DB.Where("qr_code = ?", qrCode).First(&existingOrder).Error; err != nil {
		var sampleProduct models.Product
		if err := DB.First(&sampleProduct).Error; err == nil {
			preOrder := models.Order{
				OrderType:   "pre_order",
				Status:      "pending",
				QrCode:      &qrCode,
				TotalAmount: sampleProduct.Price,
				OrderItems: []models.OrderItem{
					{
						ProductID:      sampleProduct.ID,
						Quantity:       1,
						PriceSnapshot:  sampleProduct.Price,
						MarginSnapshot: sampleProduct.SchoolMargin,
					},
				},
			}
			DB.Create(&preOrder)
			log.Println(" -> Dummy Pre-Order berhasil dibuat: ORD-PRE-20260923-001")
		}
	}
}

File: seeders/seeder.js
require('dotenv').config();
const bcrypt = require('bcryptjs');
const { sequelize, User, Product } = require('../models');

const seedData = async () => {
  try {
    console.log('Menghubungkan ke database MySQL untuk seeding...');
    await sequelize.authenticate();
    await sequelize.sync({ alter: true });

    console.log('Memeriksa dan membuat akun bawaan demo...');
    const defaultPassword = 'password123';
    const salt = await bcrypt.genSalt(10);
    const passwordHash = await bcrypt.hash(defaultPassword, salt);

    // 1. Akun Admin
    const [adminUser] = await User.findOrCreate({
      where: { nisn_nip: '9999999999' },
      defaults: {
        name: 'Administrator SIPeKa',
        password_hash: passwordHash,
        role: 'admin',
        is_active: true
      }
    });
    console.log(`✓ Admin User siap: ${adminUser.name} (${adminUser.nisn_nip})`);

    // 2. Akun Kasir
    const [kasirUser] = await User.findOrCreate({
      where: { nisn_nip: '1234567890' },
      defaults: {
        name: 'Ahmad Kasir',
        password_hash: passwordHash,
        role: 'kasir',
        is_active: true
      }
    });
    console.log(`✓ Kasir User siap: ${kasirUser.name} (${kasirUser.nisn_nip})`);

    // 3. Akun Penitip
    const [penitipUser] = await User.findOrCreate({
      where: { nisn_nip: '1122334455' },
      defaults: {
        name: 'Ibu Siti Penitip',
        password_hash: passwordHash,
        role: 'penitip',
        is_active: true
      }
    });
    console.log(`✓ Penitip User siap: ${penitipUser.name} (${penitipUser.nisn_nip})`);

    // 4. Akun Pembeli
    const [pembeliUser] = await User.findOrCreate({
      where: { nisn_nip: '3344556677' },
      defaults: {
        name: 'Budi Siswa',
        password_hash: passwordHash,
        role: 'pembeli',
        is_active: true
      }
    });
    console.log(`✓ Pembeli User siap: ${pembeliUser.name} (${pembeliUser.nisn_nip})`);

    // 5. Produk Konsinyasi Awal
    const sampleProducts = [
      {
        name: 'Roti Bakar Manis',
        price: 17500.00,
        school_margin: 1000.00,
        stock: 25,
        penitip_id: penitipUser.id
      },
      {
        name: 'Donat Cokelat Keju',
        price: 6000.00,
        school_margin: 500.00,
        stock: 30,
        penitip_id: penitipUser.id
      },
      {
        name: 'Risoles Ayam Sayur',
        price: 4000.00,
        school_margin: 500.00,
        stock: 20,
        penitip_id: penitipUser.id
      },
      {
        name: 'Es Teh Manis Segar',
        price: 3500.00,
        school_margin: 500.00,
        stock: 40,
        penitip_id: penitipUser.id
      }
    ];

    for (const prod of sampleProducts) {
      const existingProduct = await Product.findOne({
        where: { name: prod.name, penitip_id: prod.penitip_id }
      });

      if (!existingProduct) {
        await Product.create(prod);
        console.log(`✓ Produk '${prod.name}' berhasil ditambahkan ke katalog.`);
      } else {
        console.log(`- Produk '${prod.name}' sudah ada di katalog.`);
      }
    }

    console.log('\n===========================================');
    console.log('✓ Seeding data demo selesai dengan sukses!');
    console.log('===========================================');
    console.log('Kredensial Demo (Password semua akun: password123)');
    console.log('1. Admin   : NISN/NIP 9999999999');
    console.log('2. Kasir   : NISN/NIP 1234567890');
    console.log('3. Penitip : NISN/NIP 1122334455');
    console.log('4. Pembeli : NISN/NIP 3344556677');
    console.log('===========================================\n');

    process.exit(0);
  } catch (error) {
    console.error('✗ Terjadi kesalahan saat seeding data:', error);
    process.exit(1);
  }
};

seedData();
